qark/modules/findPending.py

```diff
@@ -21,456 +22,499 @@
+                                    else:
+                                        for f in t._fields:
+                                            if type(getattr(t, f)) is list:
+                                                for x in getattr(t, f):
+                                                    if type(x) is m.MethodInvocation:
+                                                        if str(x.name) == "setClass":
+                                                            explicit = True
+                                                        elif str(x.name) == "setPackage":
+                                                            explicit = True
+                                else:
+                                    break
+    else:
+        common.logger.debug("NO TYPE DECLARATIONS")
+    if not explicit:
+        if found:
+            issue = ReportIssue()
+            issue.setCategory(ExploitType.INTENT)
+            issue.setDetails("Implicit Intent: " + str(
+                intent) + " used to create instance of PendingIntent. A malicious application could potentially intercept, redirect and/or modify (in a limited manner) this Intent. Pending Intents retain the UID of your application and all related permissions, allowing another application to act as yours.  File: " + str(
+                current_file) + " More details: https://www.securecoding.cert.org/confluence/display/android/DRD21-J.+Always+pass+explicit+intents+to+a+PendingIntent")
+            issue.setFile(current_file)
+            issue.setSeverity(Severity.VULNERABILITY)
+            results.append(issue)
+
+            issue = terminalPrint()
+            issue.setLevel(Severity.VULNERABILITY)
+            issue.setData("Implicit Intent: " + str(
+                intent) + " used to create instance of PendingIntent. A malicious application could potentially intercept, redirect and/or modify (in a limited manner) this Intent. Pending Intents retain the UID of your application and all related permissions, allowing another application to act as yours.  File: " + str(
+                current_file) + " More details: https://www.securecoding.cert.org/confluence/display/android/DRD21-J.+Always+pass+explicit+intents+to+a+PendingIntent")
+            results.append(issue)
+        else:
+            # TO DO - ensure we can resolve custom intents to avoid this error
+            common.logger.debug("ERROR: COULDN'T FIND THE INTENT: " + str(
+                intent) + " This may either be due to a custom Intent class or other error")
+    return
+
+
+def intent_digger(t, intent):
+    """ Traverse the intent declaration hierarchy """
+    results = [False, False]
+    if hasattr(t, "_fields"):
+        for f in t._fields:
+            current = getattr(t, f)
+            if type(current) is list:
+                for c in current:
+                    if type(c) is m.VariableDeclaration:
```

Comment:
In general, keep in mind how many levels deep your function is going (indentation is a good measure)
if you're 8 levels in, it's going to be very difficult to hold everything in your memory, and will be more bug-prone and harder to reason about